### PR TITLE
Add update_window_auto to update window using heuristic

### PR DESCRIPTION
**Pull request by tony612**
This uses a heuristic from [a ruby http2 lib](https://github.com/igrigorik/http-2/blob/master/lib/http/2/flow_buffer.rb#L29) and nghttp.

The function name `update_window_auto` seems not very good, but I couldn't come up with a better one now 😳. I have two more PRs of cowboy and gun to use this, but will wait for this merged.

### CODE REVIEW

**Comment by essen** [comment]
Would be better just being a function calculating the optimal window update needed I think, and let cowboy/gun call it when it wants. Also I can't merge this without seeing test cases or at least the accompanying PRs.

**Comment by tony612** [comment]
@essen Do you mean a function in cowlib? Like
```
% src/cow_http2_machine.erl
cal_update_window(State=#http2_machine{remote_window=RemoteWindow}) ->
	Size = % calculate size ...
	Size.

cal_update_window(StreamID, State) ->
	Size = % calculate size ...
	Size.
```
If so, do you have any suggestion about the function naming?

I didn't add tests because I saw no tests in cowlib now. 😂

**Comment by essen** [comment]
Yeah but I mean this code is used in Cowboy and/or Gun so would need to see PRs of those at least to see how it works. Tests are also rather in those projects (and mainly in Cowboy at the moment).

And yeah that's what I mean. The reason for such a function is because there might be other requirements to apply (for example maximum window from configuration, or some kind of external backpressure), so we can't just use the value given by cow_http2_machine, we might want a lower value.

About the name, I'll think about it...

**Comment by tony612** [comment]
Another solution I thought of is letting gun/cowboy(or even users' code) provides a function to calculate the window so users can use more flexible flow control. But this will need access to `http2_machine` record.

Okay, I understand the tests thing.

**Comment by essen** [comment]
I don't think the window should be available outside Cowboy/Gun. There it should be more something like `{active,N}` ie number of messages rather than byte lengths.

**Comment by essen** [comment]
I'll describe my thoughts about that in the relevant ticket #24.

**Comment by essen** [comment]
Done. https://github.com/ninenines/gun/issues/24#issuecomment-515778827

So it's not directly related to this ticket I believe.

File: src/cow_http2_machine.erl [hunk 1]
```diff
@@ -1293,6 +1295,25 @@ update_window(StreamID, Size, State)
 	Stream = #stream{remote_window=RemoteWindow} = stream_get(StreamID, State),
 	stream_store(Stream#stream{remote_window=RemoteWindow + Size}, State).
 
+cal_increasing_window(#http2_machine{remote_window=RemoteWindow, local_settings=#{initial_window_size := WindowMaxSize}})
+		when RemoteWindow =< (WindowMaxSize / 2) ->
```
Comment (essen) [inline src/cow_http2_machine.erl]: `/` is a float operation, it's probably better to use `div`. Don't think the rounding matters much here.

**Comment by essen** [comment]
FYI https://github.com/ninenines/gun/pull/205 has a start of the flow control implementation. For HTTP/2 it basically just delays the stream window updates until flow resumes.

**Comment by essen** [comment]
Your original issue has been fixed via ninenines/cowboy#1400 and ninenines/cowlib#87. Closing, thanks!

**Pull request closed without merging**